src/Simulation/Simulators/ToffoliSimulator/Processor.cs

```diff
@@ -0,0 +1,429 @@
+                throw new InvalidOperationException($"Both input arrays for Exp (bases, qubits), must be of same size.");
+            }
+
+            foreach (var (qubit, pauli) in qubits.Zip(bases, (q, p) => (q, p))) {
+                if (CheckRotation(pauli, theta).isX) {
+                    X(qubit);
+                }
+            }
+        }
+
+        public override void ControlledExp(IQArray<Qubit> controls, IQArray<Pauli> bases, double theta, IQArray<Qubit> qubits) {
+            if (qubits.Length != bases.Length) {
+                throw new InvalidOperationException($"Both input arrays for Exp (bases, qubits), must be of same size.");
+            }
+
+            foreach (var (qubit, pauli) in qubits.Zip(bases, (q, p) => (q, p))) {
+                if (!CheckRotation(pauli, theta).safe) {
+                   throw new InvalidOperationException($"The Toffoli simulator can only perform controlled rotations of multiples of 2*pi.");
+                }
+            }
+        }
+
+        public override void ExpFrac(IQArray<Pauli> bases, long numerator, long power, IQArray<Qubit> qubits) {
+            if (qubits.Length != bases.Length) {
+                throw new InvalidOperationException($"Both input arrays for Exp (bases, qubits), must be of same size.");
+            }
+
+            foreach (var (qubit, pauli) in qubits.Zip(bases, (q, p) => (q, p))) {
+                if (CheckRotation(pauli, numerator, power).isX) {
+                    X(qubit);
+                }
+            }
+        }
+
+        public override void ControlledExpFrac(IQArray<Qubit> controls, IQArray<Pauli> bases, long numerator, long power, IQArray<Qubit> qubits) {
+            if (qubits.Length != bases.Length) {
+                throw new InvalidOperationException($"Both input arrays for Exp (bases, qubits), must be of same size.");
+            }
+
+            foreach (var (qubit, pauli) in qubits.Zip(bases, (q, p) => (q, p))) {
+                if (!CheckRotation(pauli, numerator, power).safe) {
+                   throw new InvalidOperationException($"The Toffoli simulator can only perform controlled rotations of multiples of 2*pi.");
+                }
+            }
+        }
+
+        // If the simulation value of qubit is true, when it's being reset, we
+        // restore it to false.
+        public override void Reset(Qubit qubit) {
+            SetValue(qubit, false);
+        }
+
+        // Measuring the qubit corresponds to translating the current simulation
+        // value into a Result value.
+        public override Result M(Qubit qubit) =>
+            GetValue(qubit).ToResult();
+
+
+        private Func<Pauli, Qubit, Qubit?> PauliFilter(string operation) =>
+            (Pauli p, Qubit q) =>
+                p switch {
+                    Pauli.PauliI => null,
+                    Pauli.PauliZ => q,
+                    _ => throw new InvalidOperationException($"{operation} on bases other than PauliZ not supported")
```

Comment:
>PauliZ [](start = 95, length = 6)

and PauliI

Reply:
`PauliI` is supported, but ignored. Only `PauliX` and `PauliY` raise an error.